### PR TITLE
[Fix #2831] add xref integration

### LINKED ISSUE #2831: Could CIDER provide an xref backend instead of hijacking `M-.` and `M-,`?
**Is your feature request related to a problem? Please describe.**

I'm frustrated with the way CIDER hijacks `M-.` in all clojure-mode buffers, even for projects where you don't connect to CIDER. I don't always want to connect to CIDER to navigate projects efficiently.

**Describe the solution you'd like**

I would love for CIDER to provide an xref backend instead of hijacking `M-.`, similar to how `clojure-lsp` does. 

My hope is that this would mean xref would present results both from CIDER and `clojure-lsp`, and other backends.

**Describe alternatives you've considered**

A simpler (though less ideal) fix would perhaps be to fall back to `xref-find-definitions` if there's no CIDER connection for the project.

**Additional context**

### PR DESCRIPTION
**Pull request by jiacai2050**
-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](../blob/master/.github/CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`eldev test`)
- [x] All code passes the linter (`eldev lint`) which is based on [`elisp-lint`](https://github.com/gonewest818/elisp-lint) and includes
  - [byte-compilation](https://www.gnu.org/software/emacs/manual/html_node/elisp/Byte-Compilation.html), [`checkdoc`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Tips.html), [check-declare](https://www.gnu.org/software/emacs/manual/html_node/elisp/Declaring-Functions.html), packaging metadata, indentation, and trailing whitespace checks.
- [x] You've updated the [changelog](../blob/master/CHANGELOG.md) (if adding/changing user-visible functionality)
- [x] You've updated the [user manual](../blob/master/doc) (if adding/changing user-visible functionality)

Thanks!

*If you're just starting out to hack on CIDER you might find this [section of its
manual][1] extremely useful.*

[1]: https://docs.cider.mx/cider/contributing/hacking.html

### CODE REVIEW

**Comment by bbatsov** [comment]
You should also mention this in the docs.

File: cider-find.el [hunk 1]
```diff
@@ -32,6 +32,13 @@
 
 (require 'thingatpt)
 
+(defcustom cider-enable-xref t
```
Comment (bbatsov) [inline cider-find.el:35]: The way things are currently implemented this defcustom will do something only before `cider-mode` is started. Not a big deal, but it should be noted in the docs at least. 

I'm also thinking that a name like `cider-use-xref` might be more consistent with the current names we use. I'm also worried that some people might be confused with CIDER's own `xref` functions, so we have to be a bit more descriptive in the docs I guess.

**Comment by bbatsov** [comment]
I'm assuming you'll also have to make the explicit keybindings for `M-.` and `M-,` conditional, if `xref` takes care of those.

File: cider-find.el (hunk 1 above)
Comment (jiacai2050) [reply cider-find.el]: Fixed.

File: doc/modules/ROOT/pages/usage/misc_features.adoc [hunk 2]
```diff
@@ -343,3 +343,7 @@ to format EDN using `clojure.tools.reader.edn`. The following commands are provi
 * `cider-format-edn-defun`
 * `cider-format-edn-region`
 * `cider-format-edn-buffer`
+
+== https://www.gnu.org/software/emacs/manual/html_node/emacs/Xref.html[Xref] integration
+
+Beginning with version 1.2.0, cider support Emacs built-in xref, which means `M-.` (xref-find-definitions) will just work as expected.
```
Comment (bbatsov) [inline doc/modules/ROOT/pages/usage/misc_features.adoc:349]: You should probably mention the configuration variable here as well.

**Comment by bbatsov** [comment]
One thing I don't quite get is how is this going to play with the `cider-mode` keybindings, as there are some explicit keybindings for `M-.` and `M-,` currently. @dgutov I'd appreciate any feedback from you as well, as I still haven't had the time to play with xref myself.

**Comment by dgutov** [comment]
Looking at the current patch, the user will have change the bindings in `cider-mode-map` for it to work. So you should probably change the bindings if `cider-use-xref` is non-nil too.

Even better, though, would be for Cider to migrate to Xref conditionally. If there is no important functionality that will be missed this way, of course.

File: cider-find.el [hunk 3]
```diff
@@ -226,5 +233,56 @@ thing at point."
...
+              (let* ((filename (nrepl-dict-get info "file"))
+                     (column (nrepl-dict-get info "column"))
+                     (line (nrepl-dict-get info "line"))
+                     (loc (xref-make-file-location filename line column)))
+                (xref-make filename loc)))
+            results)))
+
+(cl-defmethod xref-backend-apropos ((_backend (eql cider)) pattern)
+  "Find all symbols that match regexp PATTERN."
+  (user-error "xref-backend-apropos not supported"))
```
Comment (jiacai2050) [inline cider-find.el:285]: For apropos, I try this:

```
(when-let* ((ns (cider-current-ns))
            (results (cider-sync-request:apropos pattern ns nil t completion-ignore-case)))
  results)
```
results have no file/line/column infomation, so cannot new xref location here. Any ideas?

**Comment by jiacai2050** [comment]
I have implemented other xref functions, For apropos, I try this:

```
(cl-defmethod xref-backend-apropos ((_backend (eql cider)) pattern)
  "Find all symbols that match regexp PATTERN."
  (when-let* ((ns (cider-current-ns))
              (results (cider-sync-request:apropos pattern ns nil t completion-ignore-case)))
  results))
```
but results have no file/line/column infomation, so cannot new xref location here. Any ideas?

**Comment by dgutov** [comment]
> but results have no file/line/column infomation, so cannot new xref location here. Any ideas?

If *some* information is there, perhaps in a different form so that you can't reuse xref-file-location or xref-buffer-location, you can create your own xref types and instantiate them, like elisp-mode does, see `elisp--xref-make-xref` and the constructor that is calls.

**Comment by jiacai2050** [comment]
@dgutov Thanks for your guide.

I just realize I can reuse xref-buffer-location with another call to cider-var-info. 
PTAL

File: cider-find.el [hunk 4]
```diff
@@ -226,5 +233,71 @@ thing at point."
...
+       (save-excursion
+         (goto-char 0)
+         (forward-line (1- line))
+         (back-to-indentation)
+         (point))))))
+
+(cl-defmethod xref-backend-definitions ((_backend (eql cider)) var)
+  "Find definitions of VAR."
+  (cider-ensure-connected)
+  (cider-ensure-op-supported "ns-path")
```
[the remark this answers is not in the document]
Comment (jiacai2050) [reply cider-find.el:262]: I'm not quiet sure. I add this assert with reference to `cider-find-ns`

File: cider-find.el [hunk 5]
```diff
@@ -226,5 +233,71 @@ thing at point."
     (cider--find-ns kw-ns arg)
     (search-forward-regexp kw-to-find nil 'noerror)))
 
+(defun cider--xref-backend ()
+  "Used for xref integration."
+  'cider)
+
+(cl-defmethod xref-backend-identifier-at-point ((_backend (eql cider)))
+  "Return the relevant identifier at point."
+  (cider--kw-to-symbol (cider-symbol-at-point 'look-back)))
```
[the remark this answers is not in the document]
Comment (jiacai2050) [reply cider-find.el]: Fixed.

**Comment by bbatsov** [comment]
I guess that's good enough for a first version. Let's see what kind of feedback we'll get from the end users. Thanks for tackling the ticket!

**Comment by manuel-uberti** [comment]
Thanks to everyone involved in this.

**Pull request merged**